### PR TITLE
feat(traces): use `TraceWriter`, delegating the formatting of the print trace to `revm-inspectors`

### PR DESCRIPTION
**Pull request by zerosnacks**
## Motivation

Closes https://github.com/paradigmxyz/revm-inspectors/issues/14

Related PR: https://github.com/paradigmxyz/revm-inspectors/pull/151

## Solution

- Mutates the arena traces to populate new decoded fields in the `CallTrace`
- Brings back `decoder.decode` (now `extend_trace`)
- Delegates all formatting of the print to `revm-inspectors`

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -163,7 +163,9 @@ foundry-compilers = { version = "0.8.0", default-features = false }
 # no default features to avoid c-kzg
 revm = { version = "10.0.0", default-features = false }
 revm-primitives = { version = "5.0.0", default-features = false }
-revm-inspectors = { version = "0.1", features = ["serde"] }
+revm-inspectors = { git = "https://github.com/zerosnacks/evm-inspectors.git", rev = "df563b737c05eb62269384f432091a165cd76611", features = [
+    "serde",
+] }
```
Comment (zerosnacks) [inline Cargo.toml:168]: TODO: switch back @zerosnacks

File: crates/forge/bin/cmd/test/mod.rs [hunk 2]
```diff
@@ -517,7 +517,7 @@ impl TestArgs {
 
                         // Re-execute setup and deployment traces to collect identities created in
                         // setUp and constructor.
-                        for (kind, arena) in &result.traces {
+                        for (kind, arena) in &mut result.traces.clone() {
```
Comment (zerosnacks) [inline crates/forge/bin/cmd/test/mod.rs:520]: Not happy with the `cloning`

File: crates/forge/bin/cmd/test/mod.rs [hunk 3]
```diff
@@ -475,9 +475,9 @@ impl TestArgs {
 
                 // Identify addresses and decode traces.
                 let mut decoded_traces = Vec::with_capacity(result.traces.len());
-                for (kind, arena) in &result.traces {
+                for (kind, mut arena) in result.traces.clone() {
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/test/mod.rs:478]: here it should be &mut and decode separately, so that below you don't have to re-decode again in gas report

File: crates/evm/traces/src/decoder/mod.rs [hunk 4]
```diff
@@ -291,6 +292,27 @@ impl CallTraceDecoder {
         }
     }
 
+    pub async fn extend_traces(&self, traces: &mut CallTraceArena) {
```
Comment (mattsse) [inline crates/evm/traces/src/decoder/mod.rs:295]: needs docs

File: crates/forge/bin/cmd/test/mod.rs (hunk 2 above)
Comment (mattsse) [reply crates/forge/bin/cmd/test/mod.rs:520]: why is the clone even required?

Comment (zerosnacks) [reply crates/forge/bin/cmd/test/mod.rs]: not required here indeed

File: crates/forge/bin/cmd/test/mod.rs (hunk 3 above)
Comment (zerosnacks) [reply crates/forge/bin/cmd/test/mod.rs]: @DaniPopes fixed the `&mut`, not sure what you mean by decoding separately, do you mean moving 

```rust
   decoder.prefetch_signatures(arena.nodes()).await;
    decoder.extend_traces(arena.nodes_mut()).await;
```

out from `render_trace_arena`?

Or creating a `decoded_arenas` array to collect the decoded traces to be re-used by the gas report below?

File: Cargo.toml [hunk 5]
```diff
@@ -155,7 +155,9 @@ solang-parser = "=0.3.3"
 # no default features to avoid c-kzg
 revm = { version = "10.0.0", default-features = false }
 revm-primitives = { version = "5.0.0", default-features = false }
-revm-inspectors = { version = "0.2", features = ["serde"] }
+revm-inspectors = { git = "https://github.com/paradigmxyz/revm-inspectors.git", rev = "bc01e05", features = [
+    "serde",
+] }
```
Comment (zerosnacks) [inline Cargo.toml:160]: Pending release

**Comment by zerosnacks** [comment]
Marking as ready for review, pending `revm-inspectors` release as blocker for merging

File: crates/forge/bin/cmd/test/mod.rs (hunk 3 above)
Comment (DaniPopes) [reply crates/forge/bin/cmd/test/mod.rs]: The first, `render_trace_arena` should just be `&Arena -> String`

File: crates/evm/traces/src/decoder/mod.rs [hunk 6]
```diff
@@ -291,6 +292,31 @@ impl CallTraceDecoder {
         }
     }
 
+    /// Extends the traces with decoded data by mutating the
+    /// [CallTrace] in place. See [CallTraceDecoder::decode_function] and
+    /// [CallTraceDecoder::decode_event] for more details.
+    pub async fn extend_traces(&self, traces: &mut Vec<CallTraceNode>) {
```
Comment (DaniPopes) [inline crates/evm/traces/src/decoder/mod.rs:298]: this should be called decode_ or populate_

File: crates/evm/traces/src/decoder/mod.rs [hunk 7]
```diff
@@ -533,23 +559,23 @@ impl CallTraceDecoder {
     }
 
     /// Decodes an event.
-    pub async fn decode_event<'a>(&self, log: &'a LogData) -> DecodedCallLog<'a> {
-        let &[t0, ..] = log.topics() else { return DecodedCallLog::Raw(log) };
+    pub async fn decode_event<'a>(&self, log: &'a CallLog) -> DecodedCallLog<'a> {
```
Comment (DaniPopes) [inline crates/evm/traces/src/decoder/mod.rs:562]: why was this changed?

Comment (zerosnacks) [reply crates/evm/traces/src/decoder/mod.rs]: good point, changed it back so it again only passes down what is required i.e. `LogData`, not the entire `CallLog` that includes the decoded fields

File: crates/evm/traces/src/lib.rs [hunk 8]
```diff
@@ -59,193 +48,24 @@ pub enum DecodedCallLog<'a> {
...
-                .iter()
-                .map(|(name, value)| format!("{name}: {value}"))
-                .collect::<Vec<String>>()
-                .join(", ");
-
-            write!(s, "emit {}({params})", name.cyan())?;
-        }
-    }
+/// Render a collection of call traces.
+pub async fn render_trace_arena(arena: &mut CallTraceArena) -> Result<String, std::fmt::Error> {
```
Comment (DaniPopes) [inline crates/evm/traces/src/lib.rs:65]: unnecessary async and result

Comment (zerosnacks) [reply crates/evm/traces/src/lib.rs]: good call, updated to be not async

decoded_traces's `join_all` should still be concurrent with the changes

File: crates/evm/traces/src/decoder/mod.rs [hunk 9]
```diff
@@ -291,6 +291,31 @@ impl CallTraceDecoder {
...
 
+    /// Populates the traces with decoded data by mutating the
+    /// [CallTrace] in place. See [CallTraceDecoder::decode_function] and
+    /// [CallTraceDecoder::decode_event] for more details.
+    pub async fn populate_traces(&self, traces: &mut Vec<CallTraceNode>) {
+        for node in traces {
+            let decoded = self.decode_function(&node.trace).await;
+            node.trace.decoded.label = decoded.label;
+            node.trace.decoded.call_data = decoded.func;
+            node.trace.decoded.return_data = decoded.return_data;
```
Comment (DaniPopes) [inline crates/evm/traces/src/decoder/mod.rs:302]: Can we unify the types here? Both function and event below, I'm OK with the methods returning the type and then setting it here

File: crates/forge/bin/cmd/test/mod.rs [hunk 10]
```diff
@@ -469,7 +471,8 @@ impl TestArgs {
 
                 // Identify addresses and decode traces.
                 let mut decoded_traces = Vec::with_capacity(result.traces.len());
-                for (kind, arena) in &result.traces {
+
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/test/mod.rs:474]:
```suggestion
```

File: crates/script/src/execute.rs [hunk 11]
```diff
@@ -420,15 +421,16 @@ impl PreSimulationState {
             }
 
             shell::println("Traces:")?;
-            for (kind, trace) in &result.traces {
+            for (kind, mut trace) in result.traces.clone() {
```
Comment (DaniPopes) [inline crates/script/src/execute.rs:424]: can move the clone to inside of `should_include`

File: crates/script/src/execute.rs [hunk 12]
```diff
@@ -420,15 +421,16 @@ impl PreSimulationState {
...
                 let should_include = match kind {
```
Comment (DaniPopes) [inline crates/script/src/execute.rs:425]: (follow-up) we need to unify this logic of verbosity -> should_include between cast/test/coverage/script...

File: crates/forge/tests/it/config.rs [hunk 13]
```diff
@@ -65,23 +65,27 @@ impl TestConfig {
...
         }
         for (_, SuiteResult { test_results, .. }) in suite_result {
-            for (test_name, result) in test_results {
+            for (test_name, mut result) in test_results {
                 if self.should_fail && (result.status == TestStatus::Success) ||
                     !self.should_fail && (result.status == TestStatus::Failure)
                 {
                     let logs = decode_console_logs(&result.logs);
                     let outcome = if self.should_fail { "fail" } else { "pass" };
                     let call_trace_decoder = CallTraceDecoderBuilder::default().build();
```
Comment (DaniPopes) [inline crates/forge/tests/it/config.rs:74]: not related to this PR but I think we can decode traces using `self.known_contracts` here because currently this does nothing

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/evm/traces/src/decoder/mod.rs (hunk 9 above)
Comment (zerosnacks) [reply crates/evm/traces/src/decoder/mod.rs]: Updated to have a unified `DecodedItem` type

File: crates/script/src/execute.rs (hunk 12 above)
Comment (zerosnacks) [reply crates/script/src/execute.rs]: Opened follow-up ticket here: https://github.com/foundry-rs/foundry/issues/8339

File: crates/evm/traces/src/decoder/mod.rs (hunk 9 above)
Comment (DaniPopes) [reply crates/evm/traces/src/decoder/mod.rs]: Sorry that is not what I meant. I meant we should just reuse the same types from revm-inspectors

File: into
Comment (DaniPopes) [inline into:1]: rm

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/evm/traces/src/decoder/mod.rs (hunk 9 above)
Comment (zerosnacks) [reply crates/evm/traces/src/decoder/mod.rs]: Ah yeah that makes a lot more sense lol

**Review by DaniPopes — APPROVED** [review]

**Review by mattsse — APPROVED** [review]
